File: tries/implement_trie.py
# https://leetcode.com/problems/implement-trie-prefix-tree/


class Node:
    def __init__(self):
        self.child = {}
        self.end = False


class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        curr = self.root

        for c in word:
            if c not in curr.child:
                curr.child[c] = Node()
            curr = curr.child[c]
        curr.end = True

    def search(self, word: str) -> bool:
        curr = self.root

        for c in word:
            if c not in curr.child:
                return False
            curr = curr.child[c]
        return curr.end

    def startsWith(self, prefix: str) -> bool:
        curr = self.root

        for c in prefix:
            if c not in curr.child:
                return False
            curr = curr.child[c]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: tries/ImplementTrie.java
//https://leetcode.com/problems/implement-trie-prefix-tree/description/
public class ImplementTrie {
    public class Node {
        HashMap<Character, Node> child = new HashMap<>();
        boolean end = false;
    }

    class Trie {
        private Node root;

        public Trie() {
            root = new Node();
        }

        public void insert(String word) {
            Node cur = root;
            for (char c : word.toCharArray()) {
                cur.child.putIfAbsent(c, new Node());
                cur = cur.child.get(c);
            }
            cur.end = true;
        }

        public boolean search(String word) {
            Node cur = root;
            for (char c : word.toCharArray()) {
                if (!cur.child.containsKey(c)) {
                    return false;
                }
                cur = cur.child.get(c);
            }
            return cur.end;
        }

        public boolean startsWith(String prefix) {
            Node cur = root;
            for (char c : prefix.toCharArray()) {
                if (!cur.child.containsKey(c)) {
                    return false;
                }
                cur = cur.child.get(c);
            }
            return true;
        }
    }
}
